src/backend/cpu/kernel/copy.hpp

```diff
@@ -85,5 +86,93 @@ void copy(Array<OutT> dst, Array<InT> const src, OutT default_value, double fact
+{
+    static void copy(Array<T> dst, Array<T> const src)
+    {
+        af::dim4 src_dims       = src.dims();
+        af::dim4 dst_dims       = dst.dims();
+        af::dim4 src_strides    = src.strides();
+        af::dim4 dst_strides    = dst.strides();
+
+        T const * const src_ptr = src.get();
+        T * dst_ptr = dst.get();
+
+        // find the major-most dimension, which is linear in both arrays
+        int linear_end = 0;
+        dim_t count = 1;
+        while (linear_end < 4
+                && count == src_strides[linear_end]
+                && count == dst_strides[linear_end]) {
+            ++linear_end;
+            count *= src_dims[linear_end];
+        }
+
+        // traverse through the array using strides only until neccessary
+        if (linear_end == 4) {
+            std::memcpy(dst_ptr, src_ptr, sizeof(T) * src.elements());
+
+        } else {
+            for(dim_t l=0; l<dst_dims[3]; ++l) {
+                dim_t src_loff = l*src_strides[3];
+                dim_t dst_loff = l*dst_strides[3];
+
+                if (linear_end == 3) {
+                    dim_t dst_idx = dst_loff;
+                    dim_t src_idx = src_loff;
+                    std::memcpy(dst_ptr + dst_idx, src_ptr + src_idx, sizeof(T) * src_strides[3]);
+
+                } else {
+                    for(dim_t k=0; k<dst_dims[2]; ++k) {
+                        dim_t src_koff = k*src_strides[2];
+                        dim_t dst_koff = k*dst_strides[2];
+
+                        if (linear_end == 2) {
+                            dim_t dst_idx = dst_koff + dst_loff;
+                            dim_t src_idx = src_koff + src_loff;
+                            std::memcpy(dst_ptr + dst_idx, src_ptr + src_idx, sizeof(T) * src_strides[2]);
+
+                        } else {
+                            for(dim_t j=0; j<dst_dims[1]; ++j) {
+                                dim_t src_joff = j*src_strides[1];
+                                dim_t dst_joff = j*dst_strides[1];
+
+                                if (linear_end == 1) {
+                                    dim_t dst_idx = dst_joff + dst_koff + dst_loff;
+                                    dim_t src_idx = src_joff + src_koff + src_loff;
+                                    std::memcpy(dst_ptr + dst_idx, src_ptr + src_idx, sizeof(T) * src_strides[1]);
+
+                                } else {
+                                    for(dim_t i=0; i<dst_dims[0]; ++i) {
+                                        dim_t dst_idx = i*dst_strides[0] + dst_joff + dst_koff + dst_loff;
+                                        dim_t src_idx = i*src_strides[0] + src_joff + src_koff + src_loff;
+                                        dst_ptr[dst_idx] = src_ptr[src_idx];
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+};
```

Comment:
This looks very nice and elegant way to handle this! But the code can be simplified a bit. Can you look at what stridedCopy is doing and implement something similar? That way the size of the function can be greatly reduced.